plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -292,9 +293,52 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          }
       };
 
+      bool on_sync_block(const signed_block_ptr& block) {
+         auto& chain = chain_plug->chain();
+
+         const auto& id = block->calculate_id();
+         auto blk_num = block->block_num();
+
+         fc_dlog(_log, "syncing blockvault block ${n} ${id}", ("n", blk_num)("id", id));
+
+         // start processing of block
+         auto bsf = chain.create_block_state_future( id, block );
+
+         // abort the pending block
+         _unapplied_transactions.add_aborted( chain.abort_block() );
+
+         // push the new block
+         auto handle_error = [&](const auto& e)
+         {
+            elog((e.to_detail_string()));
+            throw;
+         };
+
+         try {
+            block_state_ptr blk_state = chain.push_block( bsf, [this]( const branch_type& forked_branch ) {
+               _unapplied_transactions.add_forked( forked_branch );
+            }, [this]( const transaction_id_type& id ) {
+               return _unapplied_transactions.get_trx( id );
+            } );
+         } catch ( const guard_exception& e ) {
+            chain_plugin::handle_guard_exception(e);
+            return false;
+         } catch ( const std::bad_alloc& ) {
+            chain_plugin::handle_bad_alloc();
+         } catch ( boost::interprocess::bad_alloc& ) {
+            chain_plugin::handle_db_exhaustion();
+         } catch( const fc::exception& e ) {
+            handle_error(e);
```

Comment:
I think we know there will be cases where blockvault may deliver an unlinkable block.  Consider a timeline with a fork, one fork wins and a snapshot is created at a blockheight that existed on both forks:

```
         LIB 
          \/
Common -> A1 -> A2 -> A3 -> A4 -> HEAD
      \-> B1 -> B2 (orphan)
```

A snapshot taken at `A1` will prune `A1`, `B1` and all their ancestors from Blockvault but not `B2` as blockvault does not reconstruct the tree and therefore cannot realize that `B2` is orphaned.  This means a `sync` can contain `B2` (and perhaps additional blocks in the `B` chain) that are unlinkable after loading the snapshot. 

After the sync, we should still be in the right place even with the erroneous unlinkable blocks.

This is a long-winded way of saying we should either pre-check that the `prev` is a known block and early out if it is not OR we should log and drop the unlinkable block exception instead of throwing out.

Reply:
This solution would solve the side effect of these erroneous unlinkable blocks, but it is really papering over the root cause (we are incorrectly or inconsistently applying pruning).

My worry is that it will try to sync now and get a lot of legitimate unlinkable blocks because of either a bug in our implementation or an inconsistent DB on the users side.  This would continue to drop everything on the floor and appear that it is fine at the end, but in reality in only applied a small number of the blocks that it needed to.

Reply:
That was how we defined pruning for blockvault so, while it is a root cause it is also expected behavior.  

To do better we either need to replicate forkdb logic in the backends/SQL which will complicate their nominal operation so that we have cleaner output during (hopefully) exceptional syncing cases where snapshots have sliced branches.  

As an alternative, we can reconstruct the trees on the client side but for the cases you are describing we would gain no additional guarantees and functionally we'd be duplicating fork-db.

Reply:
Yeah, I know that it is expected behavior given how we defined the rules for pruning (looking back at my post the work "incorrectly" shouldn't have been used).  I meant it will produce inconsistencies from being able to reapply everything that block vault owns.

And I'm not stating that we need to replicate the entirety of forkdb on the non-eos side.

But, some degree of being able to determine legitimate vs. illegitimate errors would allow for the system to not mask over or hide the original error.  Because, as you have already mentioned if the blocks are illegitimate or orphaned then we will ultimately be in a consistent state after syncing.  By creating the blanket that says any unlinkable block is acceptable, we lose that tight guarantee.  This doesn't necessarily need to be done as part of this fix (I understand the needs of this fix for today).  But, I think it would be good to circle back to this very shortly after.

Reply:
auto previous = chain.fetchblock_by_id(block->previous);

Reply:
Dlog and short circuit without failing for now.  We can revisit a better solution later as we don’t have a concrete proposal for anything that improves upon this yet.